File: algorithms/2_medium/C/01109_corporate_flight_bookings/prefix.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
__attribute__((no_sanitize("all")))
int* corpFlightBookings(int** bookings, const int rows, const int* cols_data, const int n, int* return_len)
{
    assert(n >= 0 && rows > 0 && cols_data != NULL && bookings != NULL);
    int* prefix = calloc((size_t)n + 1, sizeof(*prefix));
    assert(prefix != NULL);
    *return_len = n;

    for (int i = 0; i < rows; i += 1) {
        const int* b = bookings[i];
        assert(cols_data[i] == 3);
        int lo = b[0];
        int hi = b[1];
        int seats = b[2];
        assert(lo >= 1 && lo <= hi && hi <= n && seats >= 1);
        prefix[lo - 1] += seats;
        prefix[hi] -= seats;
    }
    for (int i = 1; i < n; i += 1) {
        prefix[i] += prefix[i - 1];
    }

    return prefix;
}

File: algorithms/2_medium/C/01109_corporate_flight_bookings/prefix.rs
impl Solution {
    pub fn corp_flight_bookings(bookings: Vec<Vec<i32>>, n: i32) -> Vec<i32> {
        assert!(n >= 0);
        let mut prefix = vec![0_i32; (n + 1) as usize];
        for b in bookings {
            debug_assert!(b.len() == 3);
            let (lo, hi, seats) = (b[0], b[1], b[2]);
            debug_assert!(lo >= 1 && lo <= hi && hi <= n && seats >= 1);
            prefix[(lo - 1) as usize] += seats;
            prefix[hi as usize] -= seats;
        }
        prefix.pop();
        for i in 1..(n as usize) {
            prefix[i] += prefix[i - 1];
        }

        prefix
    }
}
